designs/2024-config-lookup-from-file/README.md

```diff
@@ -0,0 +1,134 @@
+
+As a result, it became obvious that the modified configuration lookup strategy in flat config would not work in the long term, and we need to go back to looking up configuration files from the file being linted.
+
+## Detailed Design
+
+This proposal consists of the following changes:
+
+1. Create a new `ConfigLoader` class that manages the lookup and caching of configurations
+1. Create a new `LegacyConfigLoader` class with the same interface as `ConfigLoader` to encapsulate the current configuration lookup strategy (look up from the cwd)
+1. Use a feature flag to switch between the two modes in the `ESLint` class
+
+**Compatibility:** This proposal requires the flat config system and cannot be used with the eslintrc config system.
+
+### The `ConfigLoader` interface
+
+The following interface will be implemented twice, once as `ConfigLoader` to encapsulate the new behavior and once as `LegacyConfigLoader` to encapsulate the current behavior. This will allow the `ESLint` class to switch between these two modes easily and remove all configuration lookup-related functionality from the `ESLint` class.
+
+```ts
+interface ConfigLoader {
+
+    /**
+     * Searches the file system for the right config file to use based on the
+     * absolute path.
+     */
+    findConfigFileForPath(fileOrDirPath): Promise<string|undefined>;
+
+    /**
+     * An asynchronous call that searches for a config file from the given
+     * absolute path and returns the config array for that path.
+     */
+    loadConfigArrayForPath(fileOrDirPath): Promise<FlatConfigArray>;
+
+}
+```
+
+#### The `findConfigFileForPath()` Method
+
+This method returns the path to the config file for a given file path. When used in `LegacyConfigLoader`, this method would search from the cwd and ignore the argument that was passed in. This replaces the [`findFlatConfigFile()` method](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L266-L271) that is currently in `lib/eslint/eslint.js`.
+
+#### The `loadConfigArrayForPath()` Method
+
+This method behaves similarly as the current [`ESLint#calculateConfigForPath()` method](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L1165-L1174) except that it returns the `FlatConfigArray` instead of the config for the given file. All of the logic in `ESLint#calculateConfigForPath()` will be moved to the `ConfigLoader` class and the `ESLint` class will call the `ConfigLoader` method to provide this functionality. This requires moving the [`calculateConfigArray()` function](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L369) into `ConfigLoader` (as a private method).
```

Comment:
The current method is called `calculateConfigForFile`.

```suggestion
This method behaves similarly as the current [`ESLint#calculateConfigForFile()` method](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L1165-L1174) except that it returns the `FlatConfigArray` instead of the config for the given file. All of the logic in `ESLint#calculateConfigForFile()` will be moved to the `ConfigLoader` class and the `ESLint` class will call the `ConfigLoader` method to provide this functionality. This requires moving the [`calculateConfigArray()` function](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L369) into `ConfigLoader` (as a private method).
```